Create rule S1143[kotlin]: Jump statements should not occur in "final… (#319)
SonarSource/rspec: rules/S1143/kotlin/rule.adoc

> Using `return`, `break`, `throw`, and so on from a `finally` block suppresses the propagation of any unhandled `Throwable` which was thrown in the `try` or `catch` block.

Reviewer: ```suggestion
Using `break`, `continue`, `return` and `throw` inside of a `finally` block suppresses the propagation of any unhandled `Throwable` thrown in the `try` or `catch` block.
```

> This rule raises an issue when a jump statement (`break`, `continue`, `return`, `throw`) would force control flow to leave a `finally` block. 

Author: The original description mentions `goto` as well, I would rather write a dedicated description here.

== Noncompliant Code Example

----
fun main() {
    try {
        doSomethingWhichThrowsException(5)
        println("OK") // incorrect "OK" message is printed
    } catch (e: RuntimeException) {
        println("ERROR") // this message is not shown
    }
}

fun doSomethingWhichThrowsException(q: Int) {
    try {
        throw RuntimeException()
    } finally {
        for (i in 0..9) {
            //...
            if (q == i) {
                break // ignored
            }
        }

        /* ... */
        return  // Noncompliant - prevents the RuntimeException from being propagated
    }
}
----

== Compliant Solution

----
fun main() {
    try {
        doSomethingWhichThrowsException()
        println("OK") // incorrect "OK" message is printed
    } catch (e: RuntimeException) {

>         println("ERROR") // this message is not shown

Reviewer: I suppose we can modify these two comments to reflect the opposite in the compliant variant?

Author: Copy/Paste issue :)

    }
}

fun doSomethingWhichThrowsException(q: Int) {
    try {
        throw RuntimeException()
    } finally {
        for (i in 0..9) {
            //...
            if (q === i) {
                break // ignored
            }
        }

        /* ... */
    }
}
----

ifdef::env-github,rspecator-view[]
'''
== Comments And Links
(visible only on this page)

include::../comments-and-links.adoc[]

> endif::env-github,rspecator-view[]

Reviewer: Never sure if we want this, to me it seems like a relict from JIRA. No strong opinion, though.

Author: I think this to show comments from Jira on the RSPEC. No strong opinion here to, but try to be consistent with other languages
